enhancements/installer/aws-lb-subnet-selection.md

```diff
@@ -0,0 +1,1008 @@
+    //
+    // +optional
+    // +listType=atomic
+    // +kubebuilder:validation:XValidation:rule=`self.all(x, self.exists_one(y, x == y))`,message="subnet ids cannot contain duplicates"
+    // + Note: Though it may seem redundant, MaxItems is necessary to prevent exceeding of the cost budget for the validation rules.
+    // +kubebuilder:validation:MaxItems=10
+    IDs []AWSSubnetID `json:"ids,omitempty"`
+
+    // names specifies a list of AWS subnets by subnet name.
+    // Subnet names must not start with "subnet-", must not
+    // include commas, must be under 256 characters in length,
+    // must be unique, and the total number of subnets
+    // specified by ids and names must not exceed 10.
+    //
+    // +optional
+    // +listType=atomic
+    // +kubebuilder:validation:XValidation:rule=`self.all(x, self.exists_one(y, x == y))`,message="subnet names cannot contain duplicates"
+    // + Note: Though it may seem redundant, MaxItems is necessary to prevent exceeding of the cost budget for the validation rules.
+    // +kubebuilder:validation:MaxItems=10
+    Names []AWSSubnetName `json:"names,omitempty"`
+}
+
+// AWSSubnetID is a reference to an AWS subnet ID.
+// +kubebuilder:validation:MinLength=24
+// +kubebuilder:validation:MaxLength=24
+// +kubebuilder:validation:Pattern=`^subnet-[0-9A-Za-z]+$`
+type AWSSubnetID string
+
+// AWSSubnetName is a reference to an AWS subnet name.
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=256
+// +kubebuilder:validation:XValidation:rule=`!self.contains(',')`,message="subnet name cannot contain a comma"
+// +kubebuilder:validation:XValidation:rule=`!self.startsWith('subnet-')`,message="subnet name cannot start with 'subnet-'"
+type AWSSubnetName string
+```
+
+#### Ingress Operator Updates
+
+The Ingress Operator will be updated to consume the Ingress Config's `spec.loadBalancer.platform.aws.classicLoadBalancer.subnets`
+or `spec.loadBalancer.platform.aws.networkLoadBalancer.subnets` fields when creating IngressControllers. The
+Ingress Config `subnet` field for the corresponding load balancer type will propagate to the IngressController's relevant
+`subnets` field as long as `Subnets` haven't been defined on the IngressController (i.e. defaulting).
+
+This logic will NOT follow the existing defaulting pattern where the default gets encoded in the IngressController
+status, because the `Subnets` fields on the IngressController status represent the *actual* state. Instead of encoding
+the default in the IngressController object, it will be dynamically sourced from the Ingress Config on each reconcile.
+This makes the default `Subnets` behavior dynamic: if the subnets in the Ingress Config are updated, the
+IngressController using these defaults will also update, triggering a `Progressing` condition to notify users of the
+subnet change (described in [Effectuating Subnet Updates](/enhancements/ingress/lb-subnet-selection-aws.md#effectuating-subnet-updates)).
```

Comment:
> The issue: when defaulting Spec.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS.[Network|Classic]LoadBalancerParameters.Subnets, we are forced to also configure the required values "along the way". For example, Spec.EndpointPublishingStrategy.LoadBalancer.ProviderParameters.AWS.Type is required, so we must encode NLB or Classic, which means we are also forced to set the default in spec for this field. But that LBType default is already encoded in the status?

Was the type the only value that was needed to be added? For new clusters I guess this makes sense. But for upgrades, how would this behave, would it then populate spec, mirroring what's in the status? I guess that is probably ok. I think @deads2k had some opinion about this on the API office hours though, so maybe check with him as well.

Do you think over time that we are likely to want to enforce changing LB types? Eg blanket moving existing LBs to NLB from CLB? IIUC, this would not be something we would ever force change, and so encoding defaults in spec is in keeping with workload APIs. IIRC David seemed to think this was a configuration API though in which case what I'm saying here is not correct. Lets try and clarify what we think this is.

> Or, if we accept not to do spec defaulting this time and keep it dynamic, what do you think about making the subnets fields on the Ingress Config immutable as an alternative? 

That would certainly make things a little simpler in this regard. Have you got specific requests from PM about being able to mutate the subnets post cluster installation?

Reply:
I guess on the defaulting front. Do we need to know if a particular Ingress object had the LB type set by the person who created the Ingress vs the admin defining a cluster wide default? When would we need to understand that difference/when would an admin/end user care?

Reply:
> That's a great question. Since we'd take the cluster-wide defaults and put them in spec, we would no longer be able to distinguish between the cluster-wide defaults and the user-provided defaults. From a CIO perspective, it doesn't care where the default came from, it reconciles it the same. But from an end user perspective, they do lose the ability to distinguish whether they set the default, or the CIO chose a default for them. I don't have a concrete use case to mandate the need for distinguishing, but it is a noticeable change.

I think we want to take great consideration to this point. Perhaps having a wider conversation synchronously in the new year to understand how folks feel about this one might help us make a decision on the other parts of this defaulting story.

Everything here seems achievable, and feels like it's a direction we could take. But losing the ability to distinguish cluster wider vs individual user desire may be something we are not willing to lose

Reply:
Yea I agree, we should talk about. I'll get something scheduled for the new year.

I just thought about @deads2k's [suggestion](https://docs.google.com/document/d/1Y-Z8gaKYv5dczbdwvXanbzBl-HjZ7zasRC0WaJGBWkA/edit?tab=t.0#heading=h.4cri70x1dc7z) to create different structs in `status` to self-describe the usage. E.g. `status.globalDefaults.endpointPublishingStrategy`. That could be an approach for subnet defaulting if we wanted to keep the same patterns with spec, and possibly provide clarifications for other defaults to the users. The con is that `status` is still encoding a desired value (the default), but this might be the more acceptable compromise in this situation (vs changing the behavior of `spec` defaulting). Also, status is going to get longer, but maybe that's a just a mild inconvenience.

Reply:
I don't love the idea of adding `status.globalDefaults` as it could get quite large, redundant, and even more confusing than the status quo.  However, if you all (Grant, Joel, and David) think it's a good idea, I can go along with it.  

Would the new `globalDefaults` field include all the settings we default, or would we set fields, such as `status.globalDefaults.endpointPublishingStrategy.loadBalancer.providerParameters.aws.classicLoadBalancer.subnets`, ad hoc and leave unnecessary fields empty?  

For that matter, would we re-use the `EndpointPublishingStrategy` type and other types we use for status, or would we define parallel types with only the fields we need in `globalDefaults`?  That could get really messy.  

---

Looking at this another way, I see mentioned a few ways determine the desired subnets for an IngressController:

(a) Autodetect subnets using tags.
(b) Inherit the subnets from the cluster ingress config on creation.
(c) Inherit the subnets from the cluster ingress config on first reconciliation.  
(c) Inherit the subnets from the cluster ingress config on every reconciliation.  

(a) is the behavior without this EP, and the implied behavior with this EP when the cluster ingress config doesn't have any subnets.  We seem to be debating whether the effect of specifying subnets in the cluster config should be (b), (c), or (d).  Right?  

Would it simplify matters if we made these behaviors into an explicit choice?  That is,

```yaml
kind: IngressController
spec:
  endpointPublishingStrategy:
    type: LoadBalancerService
    loadBalancer:
      providerParameters:
        type: AWS
        aws:
          type: Classic
          classicLoadBalancer:
            # Option (a); the list must be empty.
            subnetSelection: AutodetectUsingTags
            subnets:
              ids: []
```
or
```yaml
            # Option (b); cluster-admin specifies the list on create.
            subnetSelection: ExplicitlyEnumerated
            subnets:
              ids:
              - subnet-0fcf8e0392f0910d0
              - subnet-0fcf8e0392f0910d1
```
or
```yaml
            # Option (c); controller specifies the list on first reconciliation
            # based on the cluster config.
            subnetSelection: AutomaticallyEnumeratedFromClusterConfigOnCreate
            subnets:
              ids:
              - subnet-0fcf8e0392f0910d0
              - subnet-0fcf8e0392f0910d1
```
or
```yaml
            # Option (d); the list must be empty.
            subnetSelection: EnumeratedInClusterConfig
            subnets:
              ids: []
```

Making `subnetSelection` an explicit field helps in several ways:

- No need to persist defaults to status.  
- The field specifies explicitly who or what sets the fields when.
- We can use CEL to enforce most of the restrictions we want:
  - Prohibit updating `ids` for `ExplicitlyEnumerated`.
  - Prohibit specifying nonempty `ids` for `AutodetectUsingTags`.
  - Allow empty `ids` on creation but prohibit updating `ids` once the controller has set it for `AutomaticallyEnumeratedFromClusterConfigOnCreate`.  I don't know whether we can require the field to be empty on create.  In any case, CEL cannot prevent the cluster-admin from creating the IngressController and then immediately updating `ids` before the controller gets a chance to do it.  However, I think we can live with that gap.  
  - Prohibit specifying `ids` for `EnumeratedInClusterConfig`.  Ideally, we'd reject the IngressController if `EnumeratedInClusterConfig` or `AutomaticallyEnumeratedFromClusterConfigOnCreate` were specified without specifying any subnets in the cluster config, but we can't do that with CEL; it would need to be a controller admission check.  
- For `EnumeratedInClusterConfig`, we can document that the controller will report `Progressing=True` if the cluster config changes (that is, if the service annotation doesn't match the cluster config).  

For compatibility, I figure we will default to `AutodetectUsingTags` if the cluster-admin creates an IngressController without specifying `subnetSelection`.  If the operator creates the default IngressController, it  should use `AutomaticallyEnumeratedFromClusterConfigOnCreate` if the operator observes that the cluster config specifies subnets, and `AutodetectUsingTags` otherwise.  Otherwise, the cluster-admin can choose `ExplicitlyEnumerated` or `AutomaticallyEnumeratedFromClusterConfigOnCreate`.  

The more I think about it, I don't see a good use-case for `EnumeratedInClusterConfig`.  The only benefit over `AutomaticallyEnumeratedFromClusterConfigOnCreate` is that it provides some signal if the cluster-admin updates the cluster config and forgets to update the IngressController.  However, I would lean towards omitting `EnumeratedInClusterConfig`.  

So would making `subnetSelection` an explicit field help matters?

Reply:
> The only benefit over AutomaticallyEnumeratedFromClusterConfigOnCreate is that it provides some signal if the cluster-admin updates the cluster config and forgets to update the IngressController.

With current, defaulted status fields, if the cluster config is updated, these fields get updated _unless_ there is an IngressController spec field that is set and takes precedence over the cluster config.

So I think the current behaviour of most fields is a lot like C is it not? I think what we are trying to identify here is whether we should depart from that pattern, which would then require the admin to remember, for certain cluster config defaults, that they need to go update the IngressControllers manually, which again, sounds confusing

Reply:
Thanks @Miciah for your detailed response. I think `subnetSelection` would be a verbose way of configuring subnet behavior. But I can't help but to feel that having an extra field to describe behavior of the API itself is a bit clunky. But this is something I'm trying to warm up to, as I find that APIs often have implicit behavior, usually following some sort of convention/pattern. With subnets, we have to develop a new defaulting convention/pattern, so the implicit behavior would be even less intuitive/expected.

> So I think the current behaviour of most fields is a lot like C is it not?

@JoelSpeed yea, the current behavior (e.g. like LBType) is that it gets persisted into status on first reconcile, so like C, but with status (not spec).

Maybe this is a good time to ask the question: is adding defaulting for subnets really worth the ROI? I interpreted Service Delivery's position as a "nice-to-have" defaulting. I also saw that `LBType` in Ingress Config defaults for custom IngressControllers. So nice-to-have + consistency, does it provide enough value to our users and product?

If we didn't have subnet defaulting, what would be the fallout? Users would have to copy the subnets from the `default` IngressController to their custom IngressController when they create them Day2, assuming they wanted the same config, right? Subnet defaulting doesn't enable any new functionality, it's just a UX convenience. Actually, since IC empty subnets (no-opinion) indicates Subnet Discovery, defaulting with the current `IngressControllerLB` role design actually totally prevents someone installing a cluster with explicit subnet configuration, and adding a custom ICs that uses subnet discovery (because it will always get defaulted). Maybe that's a use case to NOT default subnets. Or maybe we need separate `default` and custom IC subnet roles... 

We could rename the installer subnet role of `IngressControllerLB` to `DefaultIngressControllerLB`. That would 1. Be super explicit it only sets the `default` IngressController subnets, 2. Provide a clear path for user-created ICs subnet role to be added to the installer at a later date if it's requested (e.g. `CustomIngressControllerLB`).

So, don't take this as an effort to get myself out of doing extra work, but maybe it's someone we can punt on for the first iteration of this. If we frame IC subnet defaulting as its own RFE, would we reject it? I think punting on it would help lower the risk/scope of this proposal, as recently it has been growing.

Reply:
It does appear that the defaulting is the most complex part of this feature at the moment.

I'd suggest discussing with the team and PM about whether you really need this. I think we probably want to make a decision that it's a now or never thing though, as retro-fitting the defaulting later could be trickier, eg we would have to work out upgrade issues for existing clusters.

I could however imagine it in the future as a new field to default the non default ICs as a way to implement this, but you're right, it does exclude then, any use of subnet discovery, which we've already agreed is the majority use case.